src/HotChocolate/Neo4J/src/Data/Language/Visitor/CypherVisitor.cs

```diff
@@ -0,0 +1,122 @@
+using System.Collections.Generic;
+using HotChocolate.Language;
+
+namespace HotChocolate.Data.Neo4J.Language
+{
+    public partial class CypherVisitor
+    {
+        private readonly LinkedList<IVisitable> _currentVisitedElements = new();
+
+        /// <summary>
+        /// Target for all rendered parts of the cypher statement.
+        /// </summary>
+        private readonly CypherWriter _writer = new();
+
+        /// <summary>
+        /// Keeps track of named objects that have been already visited.
+        /// </summary>
+        private readonly HashSet<INamed> _visitedNamed = new();
+
+        /// <summary>
+        /// A set of aliased expressions that already have been seen and for which an alias must be
+        /// used on each following appearance.
+        /// </summary>
+        private readonly HashSet<AliasedExpression> _visitableToAliased = new();
+
+        /// <summary>
+        /// Keeps track if currently in an aliased expression so that the content can be skipped
+        /// when already visited.
+        /// </summary>
+        private readonly List<AliasedExpression> _currentAliasedElements = new();
+
+        /// <summary>
+        /// This keeps track on which level of the tree a separator is needed.
+        /// </summary>
+        private readonly Dictionary<int, string> _separatorOnLevel = new();
+
+        /// <summary>
+        /// Keeps track of unresolved symbolic names.
+        /// </summary>
+        private readonly Dictionary<SymbolicName, string> _resolvedSymbolicNames = new();
+
+        /// <summary>
+        /// The current level in the tree of cypher elements.
+        /// </summary>
+        private int _currentLevel;
```

Comment:
What you could do to avoid the `++`, `+1` etc. is to introduce a Scope.
Like in the filter visitor. 
On a scope you have everything that is related to this scope